internal/passwd/dk.go

```diff
@@ -0,0 +1,127 @@
+package passwd
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"golang.org/x/crypto/argon2"
+)
+
+// ===========================================================================
+// Derived Key Algorithm
+// ===========================================================================
+
+// Argon2 constants for the derived key (dk) algorithm
+// See: https://cryptobook.nakov.com/mac-and-key-derivation/argon2
+const (
+	dkAlg  = "argon2id"        // the derived key algorithm
+	dkTime = uint32(1)         // draft RFC recommends time = 1
+	dkMem  = uint32(64 * 1024) // draft RFC recommends memory as ~64MB (or as much as possible)
+	dkProc = uint8(2)          // can be set to the number of available CPUs
+	dkSLen = 16                // the length of the salt to generate per user
+	dkKLen = uint32(32)        // the length of the derived key (32 bytes is the required key size for AES-256)
+)
+
+// Argon2 variables for the derived key (dk) algorithm
+var (
+	dkParse = regexp.MustCompile(`^\$(?P<alg>[\w\d]+)\$v=(?P<ver>\d+)\$m=(?P<mem>\d+),t=(?P<time>\d+),p=(?P<procs>\d+)\$(?P<salt>[\+\/\=a-zA-Z0-9]+)\$(?P<key>[\+\/\=a-zA-Z0-9]+)$`)
+)
+
+// CreateDerivedKey creates an encoded derived key with a random hash for the password.
+func CreateDerivedKey(password string) (_ string, err error) {
```

Comment:
Instead of having one named return in the function signature, I would just do: 

```suggestion
func CreateDerivedKey(password string) (string, error) {
```

And then line 41 will have to change to the following since `err` is not yet defined

```
if _, err := rand.Read(salt); err != nil {
```